### ISSUE TITLE
Back to Basics: Linear Regression

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2708]
Back to Basics: Linear Regression<br>
<br>
http://bit.ly/2E7bpt6<br>
<br>
Gaurav Menghani<br>
<br>
<div>
<div>
<p>Taking inspiration from <a href="http://www.allthingsdistributed.com/" target="_blank">Werner Vogel’s ‘Back to Basics’ blogposts</a>, here is one of my own posts about fundamental topics. While on a long-haul flight with no internet connectivity, having exhausted the books on my kindle, and hardly any inflight-entertainment, I decided to code up Linear Regression in Python. Let’s look at both the theory and implementation of the same.</p>

<h2 id="theory">Theory</h2>

<p>Essentially, in Linear Regression, we try to estimate a dependent variable $y$, using independent variables $x_1$, $x_2$, $x_3$, $…$, using a linear model.</p>

<p>More formally,
$y = b + W_1 x_1 + W_2 x_2 + … + W_n x_n + \epsilon$.
Where, $W$ is the weight vector, $b$ is the bias term, and $\epsilon$ is the noise in the data.</p>

<p>It can be used when there is a linear relationship between the input $X$ (input vector containing all the $x_i$, and $y$).</p>

<p>One example could be, given a vending machine’s sales of different kind of soda, predict the total profit made. Let’s say there are three kinds of soda, and for each can of that variety sold, the profit is 0.25, 0.15 and 0.20 respectively. Also, we know that there will be a fixed cost in terms of electricity and maintenance for the machine, this will be our bias, and it will be negative. Let’s say it is $100. Hence, our profit will be:</p>

<p>$y = -100 + 0.25x_1 + 0.15x_2 + 0.20x_3$.</p>

<p>The problem is usually the inverse of the above example. Given the profits made by the vending machine, and sales of different kinds of soda (i.e., several pairs of $(X_i, y_i)$), find the above equation. Which would mean being able to find $b$, and $W$. There is a closed-form solution for Linear Regression, but it is expensive to compute, especially when the number of variables is large (10s of thousands).</p>

<p>Generally in Machine Learning the following approach is taken for similar problems:</p>

<ol>
<li>Make a guess for the parameters ($b$ and $W$ in our case).</li>
  <li>Compute some sort of a loss function. Which tells you how far you are from the ideal state.</li>
  <li>Find how much you should tweak your parameters to reduce your loss.</li>
</ol>
<p><strong>Step 1</strong>
The first step is fairly easy, we just pick a random $W$ and $b$. Let’s say $\theta = (W, b)$, then $h_\theta(X_i) = b + X_i.W$. Given an $X_i$, our prediction would be $h_\theta(X_i)$.</p>

<p><strong>Step 2</strong>
For the second step, one loss function could be, the average absolute difference between the prediction and the real output. This is called the ‘L1 norm’.</p>

<p>$L_1 = \frac{1}{n}\sum_{i=1}^{n} \text{abs}(h_\theta(X_i) - y_i)$</p>

<p>L1 norm is pretty good, but for our case, we will use the average of the squared difference between the prediction and the real output. This is called the ‘L2 norm’, and is <a href="http://www.chioka.in/differences-between-the-l1-norm-and-the-l2-norm-least-absolute-deviations-and-least-squares/" target="_blank">usually preferred over L1</a>.</p>

<p>$L_2 = \frac{1}{2n}\sum_{i=1}^{n} (h_\theta(X_i) - y_i)^2$.</p>

<p><strong>Step 3</strong> We have two sets of params $b$ and $W$. Ideally, we want $L_2$ to be 0. But that would depend on the choices of these params. Initially the params are randomly chosen, but we need to tweak them so that we can minimize $L_2$.</p>

<p>For this, we follow the <a href="https://en.wikipedia.org/wiki/Stochastic_gradient_descent" target="_blank">Stochastic Gradient Descent algorithm</a>. We will compute ‘partial derivatives’ / gradient of $L_2$ with respect to each of the parameters. This will tell us the slope of the function, and using this gradient, we can adjust these params to reduce the value of the method.</p>

<p>Again,</p>

<p>$L_2 = \frac{1}{2n}\sum_{i=1}^{n} (h_\theta(X_i) - y_i)^2$.</p>

<p>Deriving w.r.t. $b$ and applying chain rule,</p>

<p>$\large \frac{\partial L}{\partial b}$ = $2 . \large\frac{1}{2n}$ $\sum_{i=1}^{n} (h_\theta(X_i) - y_i) . 1$ (Since, $\frac{\partial (h_\theta(X_i) - y_i)}{\partial b} = 1$)</p>

<p>$ \implies \large \frac{\partial L}{\partial b}$ $= \sum_{i=1}^{n} (h_\theta(X_i) - y_i)$</p>

<p>Similarly, deriving w.r.t. $W_j$ and applying chain rule,</p>

<p>$\large \frac{\partial L}{\partial W_j}$ = $\large\frac{1}{n}$ $\sum_{i=1}^{n} (h_\theta(X_i) - y_i) . X_{ij}$</p>

<p>Hence, at each iteration, the updates we will perform will be,</p>

<p>$b = b - \eta \large\frac{\partial L}{\partial b}$, and,
$W_j = W_j - \eta \large\frac{\partial L}{\partial W_j}$.</p>

<p>Where, $\eta$ is what is called the ‘learning rate’, which dictates how big of an update we will make. If we choose this to be to be small, we would make very small updates. If we set it to be a large value, then we might skip over the local minima. There are a lot of variants of SGD with different tweaks around how we make the above updates.</p>

<p>Eventually we should converge to a value of $L_2$, where the gradients will be nearly 0.</p>

<h2 id="implementation">Implementation</h2>

<p>The complete implementation with dummy data in about 100 lines is <a href="https://gist.github.com/reddragon/91c023a123b8aab4c200183173e583fc" target="_blank">here</a>. A short walkthrough is below.</p>

<p>The only two libraries that we use are <code>numpy</code> (for vector operations) and <code>matplotlib</code> (for plotting losses). We generate random data without any noise.</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
</pre></td>
<td><pre><code><span><span>def</span> <span>linear_sum</span><span>(</span><span>X</span><span>,</span> <span>W</span><span>,</span> <span>b</span><span>):</span>
</span><span>    <span>return</span> <span>X</span><span>.</span><span>dot</span><span>(</span><span>W</span><span>)</span> <span>+</span> <span>b</span>
</span><span>
</span><span><span>def</span> <span>data_gen</span><span>(</span><span>num_rows</span><span>,</span> <span>num_feats</span><span>,</span> <span>op_gen_fn</span><span>=</span><span>linear_sum</span><span>):</span>
</span><span>    <span>data</span> <span>=</span> <span>{}</span>
</span><span>    <span>W</span> <span>=</span> <span>0.1</span> <span>*</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>num_feats</span><span>)</span>
</span><span>    <span>b</span> <span>=</span> <span>0.1</span> <span>*</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>()</span>
</span><span>    <span>data</span><span>[</span><span>'X'</span><span>]</span> <span>=</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>num_rows</span><span>,</span> <span>num_feats</span><span>)</span>
</span><span>    <span>data</span><span>[</span><span>'y'</span><span>]</span> <span>=</span> <span>op_gen_fn</span><span>(</span><span>data</span><span>[</span><span>'X'</span><span>],</span> <span>W</span><span>,</span> <span>b</span><span>)</span>
</span><span>    <span>return</span> <span>data</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>Where <code>num_rows</code> is $n$ as used in the above notation, and <code>num_feats</code> is the number of variables. We define the class <code>LinearRegression</code>, where we initialize <code>W</code> and <code>b</code> randomly initially. Also, the <code>predict</code> method computes $h_\theta(X)$.</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
</pre></td>
<td><pre><code><span><span>class</span> <span>LinearRegression</span><span>(</span><span>object</span><span>):</span>
</span><span>    <span>def</span> <span>__init__</span><span>(</span><span>self</span><span>):</span>
</span><span>        <span>self</span><span>.</span><span>W</span> <span>=</span> <span>None</span>
</span><span>        <span>self</span><span>.</span><span>b</span> <span>=</span> <span>0</span>
</span><span>
</span><span>    <span>def</span> <span>init_matrix</span><span>(</span><span>self</span><span>,</span> <span>X</span><span>):</span>
</span><span>        <span>self</span><span>.</span><span>W</span> <span>=</span> <span>0.1</span> <span>*</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>(</span><span>X</span><span>.</span><span>shape</span><span>[</span><span>1</span><span>])</span>
</span><span>        <span>self</span><span>.</span><span>b</span> <span>=</span> <span>0.1</span> <span>*</span> <span>np</span><span>.</span><span>random</span><span>.</span><span>randn</span><span>()</span>
</span><span>
</span><span>    <span>def</span> <span>predict</span><span>(</span><span>self</span><span>,</span> <span>X</span><span>):</span>
</span><span>        <span>if</span> <span>self</span><span>.</span><span>W</span> <span>is</span> <span>None</span><span>:</span>
</span><span>            <span>self</span><span>.</span><span>init_matrix</span><span>(</span><span>X</span><span>)</span>
</span><span>
</span><span>        <span>return</span> <span>X</span><span>.</span><span>dot</span><span>(</span><span>self</span><span>.</span><span>W</span><span>)</span> <span>+</span> <span>self</span><span>.</span><span>b</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>The crux of the code is in the <code>train</code> method, where we compute the gradients.</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
<span>11</span>
<span>12</span>
<span>13</span>
<span>14</span>
<span>15</span>
<span>16</span>
<span>17</span>
<span>18</span>
<span>19</span>
<span>20</span>
<span>21</span>
<span>22</span>
<span>23</span>
<span>24</span>
<span>25</span>
<span>26</span>
<span>27</span>
<span>28</span>
<span>29</span>
<span>30</span>
<span>31</span>
<span>32</span>
<span>33</span>
<span>34</span>
<span>35</span>
<span>36</span>
<span>37</span>
<span>38</span>
<span>39</span>
<span>40</span>
</pre></td>
<td><pre><code><span>    <span>def</span> <span>train</span><span>(</span><span>self</span><span>,</span> <span>data</span><span>,</span> <span>iters</span><span>,</span> <span>lr</span><span>):</span>
</span><span>        <span>X</span> <span>=</span> <span>data</span><span>[</span><span>'X'</span><span>]</span>
</span><span>        <span>y</span> <span>=</span> <span>data</span><span>[</span><span>'y'</span><span>]</span>
</span><span>
</span><span>        <span>orig_pred</span> <span>=</span> <span>self</span><span>.</span><span>predict</span><span>(</span><span>X</span><span>)</span>
</span><span>        <span>orig_l1</span> <span>=</span> <span>self</span><span>.</span><span>l1_loss</span><span>(</span><span>orig_pred</span><span>,</span> <span>y</span><span>)</span>
</span><span>        <span>orig_l2</span> <span>=</span> <span>self</span><span>.</span><span>l2_loss</span><span>(</span><span>orig_pred</span><span>,</span> <span>y</span><span>)</span>
</span><span>
</span><span>        <span>l1</span> <span>=</span> <span>orig_l1</span>
</span><span>        <span>l2</span> <span>=</span> <span>orig_l2</span>
</span><span>
</span><span>        <span>l1_losses</span> <span>=</span> <span>[]</span>
</span><span>        <span>l2_losses</span> <span>=</span> <span>[]</span>
</span><span>        <span>l1_losses</span><span>.</span><span>append</span><span>(</span><span>l1</span><span>)</span>
</span><span>        <span>l2_losses</span><span>.</span><span>append</span><span>(</span><span>l2</span><span>)</span>
</span><span>
</span><span>        <span>for</span> <span>it</span> <span>in</span> <span>range</span><span>(</span><span>iters</span><span>):</span>
</span><span>            <span>pred</span> <span>=</span> <span>self</span><span>.</span><span>predict</span><span>(</span><span>X</span><span>)</span>
</span><span>
</span><span>            <span># Computing gradients</span>
</span><span>            <span>s1</span> <span>=</span> <span>(</span><span>pred</span> <span>-</span> <span>y</span><span>)</span>
</span><span>            <span>s2</span> <span>=</span> <span>np</span><span>.</span><span>multiply</span><span>(</span><span>X</span><span>,</span> <span>np</span><span>.</span><span>repeat</span><span>(</span><span>s1</span><span>,</span> <span>X</span><span>.</span><span>shape</span><span>[</span><span>1</span><span>])</span>\
</span><span>                    <span>.</span><span>reshape</span><span>(</span><span>X</span><span>.</span><span>shape</span><span>[</span><span>0</span><span>],</span> <span>X</span><span>.</span><span>shape</span><span>[</span><span>1</span><span>]))</span>
</span><span>            <span>wGrad</span> <span>=</span> <span>np</span><span>.</span><span>sum</span><span>(</span><span>s2</span><span>,</span> <span>axis</span><span>=</span><span>0</span><span>)</span> <span>/</span> <span>X</span><span>.</span><span>shape</span><span>[</span><span>0</span><span>]</span>
</span><span>            <span>bGrad</span> <span>=</span> <span>np</span><span>.</span><span>sum</span><span>(</span><span>s1</span><span>)</span> <span>*</span> <span>1.0</span> <span>/</span> <span>X</span><span>.</span><span>shape</span><span>[</span><span>0</span><span>]</span>
</span><span>
</span><span>            <span># Updates to W and b</span>
</span><span>            <span>wDelta</span> <span>=</span> <span>-</span><span>lr</span> <span>*</span> <span>wGrad</span>
</span><span>            <span>bDelta</span> <span>=</span> <span>-</span><span>lr</span> <span>*</span> <span>bGrad</span>
</span><span>            <span>self</span><span>.</span><span>W</span> <span>+=</span> <span>wDelta</span>
</span><span>            <span>self</span><span>.</span><span>b</span> <span>+=</span> <span>bDelta</span>
</span><span>
</span><span>            <span>l1</span> <span>=</span> <span>self</span><span>.</span><span>l1_loss</span><span>(</span><span>pred</span><span>,</span> <span>y</span><span>)</span>
</span><span>            <span>l2</span> <span>=</span> <span>self</span><span>.</span><span>l2_loss</span><span>(</span><span>pred</span><span>,</span> <span>y</span><span>)</span>
</span><span>
</span><span>            <span>l1_losses</span><span>.</span><span>append</span><span>(</span><span>l1</span><span>)</span>
</span><span>            <span>l2_losses</span><span>.</span><span>append</span><span>(</span><span>l2</span><span>)</span>
</span><span>
</span><span>        <span>print</span> <span>'Original L1 loss: </span><span>%Lf</span><span>, L2 loss: </span><span>%Lf</span><span> '</span> <span>%</span> <span>(</span><span>orig_l1</span><span>,</span> <span>orig_l2</span><span>)</span>
</span><span>        <span>print</span> <span>'Final L1 loss: </span><span>%Lf</span><span>, L2 loss: </span><span>%Lf</span><span>'</span> <span>%</span> <span>(</span><span>l1</span><span>,</span> <span>l2</span><span>)</span>
</span></code></pre></td>
</tr></table></div></figure></div>

<p>For the given input, with the fixed seed and five input variables, the solution as per the code is:</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
</pre></td>
<td><pre><code><span>- b:  -0.081910978042
</span><span>- W:  [-0.04354505  0.00196225 -0.2143796   0.1617485  -0.17349839]</span></code></pre></td>
</tr></table></div></figure></div>

<p>This is how the $L_2$ loss converges over number of iterations:</p>

<p><img src="http://blog.gaurav.im/images/2017/04/05/linear_reg.png" title="Convergence of L2 loss"></p>

<p>To verify that this is actually correct, I serialized the input to a CSV file and used R to solve this.</p>

<div><figure><div><table><tr>
<td><pre><span>1</span>
<span>2</span>
<span>3</span>
<span>4</span>
<span>5</span>
<span>6</span>
<span>7</span>
<span>8</span>
<span>9</span>
<span>10</span>
</pre></td>
<td><pre><code><span>&gt; df &lt;- read.delim(file="lin_reg_data", header=F, sep=",")
</span><span>&gt; lm(df[,1] ~ df[,2]+df[,3]+df[,4]+df[,5]+df[,6])
</span><span>
</span><span>Call:
</span><span>lm(formula = df[, 1] ~ df[, 2] + df[, 3] + df[, 4] + df[, 5] +
</span><span>    df[, 6])
</span><span>
</span><span>Coefficients:
</span><span>(Intercept)      df[, 2]      df[, 3]      df[, 4]      df[, 5]      df[, 6]
</span><span>  -0.084175    -0.041676    -0.005627    -0.213620     0.164027    -0.179344</span></code></pre></td>
</tr></table></div></figure></div>

<p>The <code>intercept</code> is same as $b$, and the rest five outputs are the $W_i$, and are similar to what my code found.</p>
</div>

  </div><br>
<br>
<br>
<br>
<br>
<br>
via blog.gaurav.im http://blog.gaurav.im<br>
<br>
February 12, 2019 at 11:21PM
